File: public/packages/dotnet/resilience/CircuitBreaker/CircuitBreaker.cs
// -----------------------------------------------------------------------
// <copyright file="CircuitBreaker.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Resilience.CircuitBreaker;

/// <summary>
/// Lightweight three-state circuit breaker (Closed / Open / Half-Open) for
/// protecting async operations against sustained downstream failures. Tracks
/// consecutive failures and fast-fails when a threshold is reached, avoiding
/// wasted timeout waits.
/// </summary>
/// <typeparam name="T">The return type of the protected operation.</typeparam>
/// <remarks>
/// Thread-safe via <see cref="Interlocked"/> operations — no locks required.
/// The probe-in-flight flag ensures only one Half-Open probe runs at a time;
/// concurrent callers during a probe receive the fallback (or
/// <see cref="CircuitOpenException"/> when no fallback is supplied).
/// </remarks>
public sealed class CircuitBreaker<T>
{
    private readonly Func<T, bool> r_isFailure;
    private readonly int r_failureThreshold;
    private readonly long r_cooldownMs;
    private readonly Func<long> r_now;
    private readonly Action<CircuitState, CircuitState>? r_onStateChange;

    private int _state; // CircuitState cast to int for Interlocked
    private int _failureCount;
    private long _openedAt;
    private int _probeInFlight; // 0 or 1

    /// <summary>
    /// Initializes a new instance of the <see cref="CircuitBreaker{T}"/> class.
    /// </summary>
    ///
    /// <param name="isFailure">
    /// Predicate that inspects a returned value to decide whether it counts
    /// as a failure. Thrown exceptions ALWAYS count as failures; this
    /// predicate adds value-based failures (e.g. <c>r => !r.Success</c>).
    /// </param>
    /// <param name="options">Optional configuration. Defaults applied if null.</param>
    /// <param name="onStateChange">
    /// Optional callback fired on state transitions. Invoked synchronously on
    /// the thread that triggered the transition.
    /// <para>
    /// <b>Footgun:</b> a throwing callback REPLACES the upstream exception that
    /// caused the transition. A buggy logger or metric emitter inside this
    /// callback can swap a meaningful "TimeoutException from upstream X"
    /// with its own "InvalidOperationException from logger" — making outage
    /// diagnosis painful. Wrap the callback body in your own try/catch (or
    /// keep it to plain log/metric calls that don't throw) to preserve the
    /// upstream exception for callers.
    /// </para>
    /// </param>
    public CircuitBreaker(
        Func<T, bool> isFailure,
        CircuitBreakerOptions? options = null,
        Action<CircuitState, CircuitState>? onStateChange = null)
    {
        // CircuitBreakerOptions is the single source of truth for defaults
        // — falling back to a parameterless ctor here delegates that work.
        options ??= new CircuitBreakerOptions();

        r_isFailure = isFailure;
        r_failureThreshold = options.FailureThreshold;
        r_cooldownMs = (long)options.CooldownDuration.TotalMilliseconds;
        r_now = options.NowFunc;
        r_onStateChange = onStateChange;
    }

    /// <summary>
    /// Gets the current circuit state.
    /// </summary>
    public CircuitState State => (CircuitState)Volatile.Read(ref _state);

    /// <summary>
    /// Gets the current consecutive failure count. Reset to zero on every
    /// observed success and on <see cref="Reset"/>.
    /// </summary>
    public int FailureCount => Volatile.Read(ref _failureCount);

    /// <summary>
    /// Executes an operation through the circuit breaker.
    /// </summary>
    ///
    /// <param name="operation">The async operation to protect.</param>
    /// <param name="fallback">
    /// Optional fallback invoked when the circuit is open or while a probe is
    /// already in flight.
    /// </param>
    /// <param name="ct">Cancellation token.</param>
    ///
    /// <returns>The operation result, or the fallback result when open.</returns>
    ///
    /// <exception cref="CircuitOpenException">
    /// Thrown when the circuit is open and no fallback is supplied.
    /// </exception>
    public async ValueTask<T> ExecuteAsync(
        Func<CancellationToken, ValueTask<T>> operation,
        Func<ValueTask<T>>? fallback = null,
        CancellationToken ct = default)
    {
        // Open → Half-Open transition once the cooldown has elapsed.
        if (State == CircuitState.Open)
        {
            var elapsed = r_now() - Interlocked.Read(ref _openedAt);
            if (elapsed >= r_cooldownMs)
            {
                if (Interlocked.CompareExchange(
                        ref _state,
                        (int)CircuitState.HalfOpen,
                        (int)CircuitState.Open) == (int)CircuitState.Open)
                    r_onStateChange?.Invoke(CircuitState.Open, CircuitState.HalfOpen);
            }
        }

        switch (State)
        {
            // Open: fast-fail.
            case CircuitState.Open when fallback is not null:
                return await fallback();

            case CircuitState.Open:
                throw new CircuitOpenException();

            // Half-Open: only one probe at a time; everyone else gets the fallback.
            case CircuitState.HalfOpen
                when Interlocked.CompareExchange(ref _probeInFlight, 1, 0) != 0:
            {
                if (fallback is not null)
                    return await fallback();

                throw new CircuitOpenException();
            }

            // Closed, or the Half-Open probe winner: actually run the operation.
            case CircuitState.Closed:
            default:
            {
                try
                {
                    var result = await operation(ct);

                    if (r_isFailure(result))
                    {
                        RecordFailure();
                        return result;
                    }

                    RecordSuccess();
                    return result;
                }
                catch
                {
                    RecordFailure();
                    throw;
                }
            }
        }
    }

    /// <summary>
    /// Manually resets the circuit to <see cref="CircuitState.Closed"/> and
    /// clears the failure count.
    /// </summary>
    public void Reset()
    {
        Interlocked.Exchange(ref _failureCount, 0);
        Interlocked.Exchange(ref _probeInFlight, 0);

        var prev = (CircuitState)Interlocked.Exchange(
            ref _state,
            (int)CircuitState.Closed);

        if (prev != CircuitState.Closed)
            r_onStateChange?.Invoke(prev, CircuitState.Closed);
    }

    private void RecordSuccess()
    {
        Interlocked.Exchange(ref _failureCount, 0);
        Interlocked.Exchange(ref _probeInFlight, 0);

        var prev = (CircuitState)Interlocked.Exchange(
            ref _state,
            (int)CircuitState.Closed);

        if (prev != CircuitState.Closed)
            r_onStateChange?.Invoke(prev, CircuitState.Closed);
    }

    private void RecordFailure()
    {
        var count = Interlocked.Increment(ref _failureCount);
        Interlocked.Exchange(ref _probeInFlight, 0);

        // Half-Open probe failed → straight back to Open.
        if (State == CircuitState.HalfOpen)
        {
            Interlocked.Exchange(ref _openedAt, r_now());

            var prev = (CircuitState)Interlocked.Exchange(
                ref _state,
                (int)CircuitState.Open);

            if (prev != CircuitState.Open)
                r_onStateChange?.Invoke(prev, CircuitState.Open);

            return;
        }

        // Closed → Open once the threshold is hit.
        if (count >= r_failureThreshold)
        {
            Interlocked.Exchange(ref _openedAt, r_now());

            var prev = (CircuitState)Interlocked.Exchange(
                ref _state,
                (int)CircuitState.Open);

            if (prev != CircuitState.Open)
                r_onStateChange?.Invoke(prev, CircuitState.Open);
        }
    }
}

File: public/packages/typescript/resilience/src/circuit-breaker/circuit-breaker.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import type { CircuitBreakerOptions } from "./circuit-breaker-options.js";
import { CircuitOpenError } from "./circuit-open-error.js";
import { CircuitState } from "./circuit-state.js";

/**
 * Three-state circuit breaker. Mirrors .NET `CircuitBreaker<T>`.
 * Closed → after `failureThreshold` consecutive failures → Open. Open →
 * after `cooldownMs` → HalfOpen. HalfOpen → success → Closed (resets
 * counter); HalfOpen → failure → Open (cooldown re-armed).
 *
 * HalfOpen admits exactly ONE probe at a time: the first caller to reach the
 * HalfOpen slot runs the operation; concurrent callers (who observe the probe
 * already in flight) receive the `fallback` result (when one is supplied) or a
 * {@link CircuitOpenError}. JS is single-threaded, so a plain boolean flag —
 * check-and-set synchronously BEFORE the first `await` — is the structural
 * equivalent of .NET's lock-free `Interlocked.CompareExchange` on its
 * probe-in-flight flag.
 *
 * Both thrown errors and returned values satisfying {@link CircuitBreakerOptions.isFailure}
 * count as failures (the value is still returned, not re-thrown).
 */
export class CircuitBreaker<T> {
  private state: CircuitState = CircuitState.Closed;
  private consecutiveFailures = 0;
  private openedAt = 0;
  private probeInFlight = false;
  private readonly now: () => number;

  constructor(private readonly opts: CircuitBreakerOptions<T>) {
    if (opts.failureThreshold < 1)
      throw new RangeError("failureThreshold must be ≥ 1");
    if (opts.cooldownMs < 0) throw new RangeError("cooldownMs must be ≥ 0");
    this.now = opts.nowFunc ?? Date.now;
  }

  /**
   * Current state — for tests + observability. A pure read: when the circuit
   * is Open and the cooldown has elapsed it REPORTS HalfOpen (the state the
   * next {@link execute} would transition into) without mutating or firing
   * {@link CircuitBreakerOptions.onStateChange}; the actual transition (and the
   * callback) happen inside {@link execute}.
   */
  get currentState(): CircuitState {
    if (this.state === CircuitState.Open && this.cooldownExpired())
      return CircuitState.HalfOpen;
    return this.state;
  }

  /**
   * Executes `op` through the breaker. When the circuit is Open — or HalfOpen
   * with a probe already in flight — returns `fallback()` if supplied, else
   * throws {@link CircuitOpenError}. Mirrors .NET
   * `ExecuteAsync(operation, fallback?, ct)`.
   */
  async execute(
    op: () => Promise<T>,
    fallback?: () => T | Promise<T>,
  ): Promise<T> {
    // Open → HalfOpen once the cooldown has elapsed. Done here (not in the
    // getter) so the transition mutates state AND fires onStateChange exactly
    // once, mirroring .NET's ExecuteAsync Open→HalfOpen CompareExchange.
    if (this.state === CircuitState.Open && this.cooldownExpired())
      this.transitionTo(CircuitState.HalfOpen);

    // Fast-fail when Open.
    if (this.state === CircuitState.Open) {
      if (fallback !== undefined) return fallback();
      throw new CircuitOpenError();
    }

    // HalfOpen: admit exactly one probe. The check-and-set is synchronous
    // (before any await), so concurrent callers cannot both win the slot.
    if (this.state === CircuitState.HalfOpen) {
      if (this.probeInFlight) {
        if (fallback !== undefined) return fallback();
        throw new CircuitOpenError();
      }
      this.probeInFlight = true;
    }

    // Closed, or the HalfOpen probe winner: run the operation.
    try {
      const result = await op();
      if (this.opts.isFailure?.(result) === true) {
        // A returned-but-failed value counts toward the breaker WITHOUT
        // throwing — the caller still receives the value.
        this.onFailure();
        return result;
      }
      this.onSuccess();
      return result;
    } catch (err) {
      this.onFailure();
      throw err;
    }
  }

  /**
   * Manually resets the circuit to {@link CircuitState.Closed} and clears the
   * failure count + probe flag. Fires {@link CircuitBreakerOptions.onStateChange}
   * only when the state actually changed. Mirrors .NET `Reset()`.
   */
  reset(): void {
    this.consecutiveFailures = 0;
    this.probeInFlight = false;
    this.transitionTo(CircuitState.Closed);
  }

  private cooldownExpired(): boolean {
    return this.now() - this.openedAt >= this.opts.cooldownMs;
  }

  private onSuccess(): void {
    this.consecutiveFailures = 0;
    this.probeInFlight = false;
    this.transitionTo(CircuitState.Closed);
  }

  private onFailure(): void {
    this.probeInFlight = false;
    if (this.state === CircuitState.HalfOpen) {
      // Single failure in HalfOpen re-arms the cooldown.
      this.openedAt = this.now();
      this.transitionTo(CircuitState.Open);
      return;
    }
    this.consecutiveFailures++;
    if (this.consecutiveFailures >= this.opts.failureThreshold) {
      this.openedAt = this.now();
      this.transitionTo(CircuitState.Open);
    }
  }

  /**
   * Sets the state and fires {@link CircuitBreakerOptions.onStateChange} only
   * on a REAL transition (no-op when already in `to`) — mirrors .NET's
   * `prev != newState` guard around every `onStateChange` invocation.
   */
  private transitionTo(to: CircuitState): void {
    const from = this.state;
    if (from === to) return;
    this.state = to;
    this.opts.onStateChange?.(from, to);
  }
}
